packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -2297,6 +2325,37 @@ Hardhat Network's forking functionality only works with blocks from at least spu
 
     return { maxFeePerGas, maxPriorityFeePerGas };
   }
+
+  /** search this._hardforkActivationBlocks for the highest block number that
+   * isn't higher than blockNum, and then return that found block number's
+   * associated hardfork name. */
+  private _selectHardforkFromActivations(
+    blockNum: BN
+  ): typeof HARDHAT_NETWORK_SUPPORTED_HARDFORKS[number] {
+    let highestFound: {
+      hardfork: typeof HARDHAT_NETWORK_SUPPORTED_HARDFORKS[number];
+      block: number;
+    } = { hardfork: "", block: 0 };
+    Object.entries(this._hardforkActivationBlocks).forEach((entry) => {
+      const [hardfork, block] = entry;
+      if (!HARDHAT_NETWORK_SUPPORTED_HARDFORKS.includes(hardfork)) {
+        throw new InvalidInputError(
+          `hardforkActivationBlocks contained unsupported hardfork "${hardfork}"`
+        );
+      }
+      if (block > highestFound.block && new BN(block).lte(blockNum)) {
+        highestFound = { hardfork, block };
+      }
+    });
+    if (highestFound.hardfork === "" && highestFound.block === 0) {
+      throw new InvalidJsonInputError(
+        `No known hardfork for block ${blockNum}. HardhatNode config's hardforkActivationBlocks: ${JSON.stringify(
```

Comment:
This error message needs to be a lot better.  Explain the situation, and give a link to the documentation.